[libc] Alternative algorithm for decimal FP printf

statham-arm opened the pull request:
The existing options for bin→dec float conversion are all based on the Ryū algorithm, which generates 9 output digits at a time using a table lookup. For users who can't afford the space cost of the table, the table-lookup subroutine is replaced with one that computes the needed table entry on demand, but the algorithm is otherwise unmodified.

The performance problem with computing table entries on demand is that now you need to calculate a power of 10 for each 9 digits you output. But if you're calculating a custom power of 10 anyway, it's easier to just compute one, and multiply the _whole_ mantissa by it.

This patch adds a header file alongside `float_dec_converter.h`, which replaces the whole Ryū system instead of just the table-lookup routine, implementing this alternative simpler algorithm. The result is accurate enough to satisfy (minimally) the accuracy demands of IEEE 754-2019 even in 128-bit long double. The new float128 test cases demonstrate this by testing the cases closest to the 39-digit rounding boundary.

In my tests of generating 39 output digits (the maximum number supported by this algorithm) this code is also both faster and smaller than the USE_DYADIC_FLOAT version of the existing Ryū code.

statham-arm committed:
[libc] Alternative algorithm for decimal FP printf

The existing options for bin→dec float conversion are all based on the
Ryū algorithm, which generates 9 output digits at a time using a table
lookup. For users who can't afford the space cost of the table, the
table-lookup subroutine is replaced with one that computes the needed
table entry on demand, but the algorithm is otherwise unmodified.

The performance problem with computing table entries on demand is that
now you need to calculate a power of 10 for each 9 digits you output.
But if you're calculating a custom power of 10 anyway, it's easier to
just compute one, and multiply the _whole_ mantissa by it.

This patch adds a header file alongside `float_dec_converter.h`, which
replaces the whole Ryū system instead of just the table-lookup
routine, implementing this alternative simpler algorithm. The result
is accurate enough to satisfy (minimally) the accuracy demands of IEEE
754-2019 even in 128-bit long double. The new float128 test cases
demonstrate this by testing the cases closest to the 39-digit rounding
boundary.

In my tests of generating 39 output digits (the maximum number
supported by this algorithm) this code is also both faster and smaller
than the USE_DYADIC_FLOAT version of the existing Ryū code.

statham-arm commented:
This PR has a performance dependency on #123580. I haven't listed a formal dependency between the two, because in _correctness_ terms they're independent. But the claim in my commit message here that the new USE_FLOAT320 algorithm is faster than USE_DYADIC_FLOAT is only true once IntegerToString is optimized. So I don't intend to land this until after #123580.

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+  // need to iterate.
+  fputil::DyadicFloat<DF_BITS> ten(Sign::POS, 1, 5);
+
+  while (true) {
```
I'm pretty sure there's a way to calculate this without the while loop. Specifically I think that if we cap the number of requested digits (`log10_input - log10_low_digit`) to be at most `min(MAX_DIGITS, length of number)` you'll be within one step of the true result.
I think there are some other tweaks as well. I'll leave those on the specific spots.

statham-arm replied:
I think you're probably right that we can guarantee to use only one iteration of the main algorithm, but I think it's slightly more complicated than you say, because of the need to round correctly.

For some binary exponents, there's only one choice of power of 10, because everything from 2^n to 2^(n+1) has the same number of decimal digits. Other binary exponents straddle a power-of-10 boundary. For the latter, I think what you'd have to do is to convert based on the smaller exponent, so that the decimal string is either the right length or one digit too long. But if it's one digit too long, you must make sure to repeat the rounding _from the full version_ and not just discard the final decimal digit and use it to re-round what's left. (Or, equivalently, remember whether the rounded result was smaller or larger.)

michaelrj-google replied:
That's a good point about needing to be careful about the rounding, I wasn't thinking about that at the time. The main thing I'd noticed was one of the test cases (line 1567 in `sprintf_test.cpp`) seemed to be taking an inefficient path: Since the precision was 500 it started with an exponent of -500, which is well out of the realm of possibility. I realize it skips to a precision of `MAX_DIGITS` on the next iteration, but it still seems like a lot more steps than necessary.

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+        // In E mode, just try again with the exponent one higher.
+        log10_input++;
+      } else {
+        // In F mode, if the number would generate too many digits to fit, then
+        // reset to E mode, so that we generate the most significant digits of
+        // the number, and rely on our caller to append zeroes at the end.
+        precision = MAX_DIGITS;
+        e_mode = true;
```
You can check if an `f` conversion needs to be done in `e` mode before the loop, because that will happen when the number of digits generated is more than `MAX_DIGITS` and the formula for the max number of digits in the number is `digits = exponent >= 0 ? log10_input + max(0, fract_width - exp) : log10_input - exponent`.

statham-arm replied:
I don't think that formula can be quite right, because of the thing I mention in the comment above, where some binary exponents can go with two possible decimal exponents depending on the mantissa, because a power of 10 lies in their interval. So any test that doesn't even look at the mantissa _must_ be potentially inaccurate.

But I'm sure you're right that _most_ cases of "too many output digits, go to e mode" can be detected before the conversion to decimal digits, and we can narrow it down so that the remaining ones only generate one digit too many, which can be dealt with by re-rounding the same float320 output.

michaelrj-google replied:
I think you're right about there being possibly one more digit, and I forgot to add `fract_width` to the second term, but I think if you add 1 you'd get a fairly tight upper bound to the number of digits. It's a similar formula to what is used in `float_to_string` for `length_for_num` and `zero_blocks_after_point`.

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+      // In E mode, we expect exactly the right number of output digits, so we
+      // must increment or decrement the exponent if we had too many or too
+      // few.
+      if (view.size() > size_t(precision)) { /* too big */
+        log10_input++;
+        continue;
```
In this case the output buffer contains all the correct digits, it just also has some extras. It's possible to avoid a complete recalculation by decrementing `output.ndigits` and incrementing `output.exponent` the correct number of times before returning.

statham-arm replied:
As I mention in the comment above, I think this needs more care over re-rounding than you mention, but fair enough.

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+  for (int pos = start; pos < limit; ++pos) {
+    if (pos >= 0 && pos < int(output.ndigits)) {
+      // Fetch a digit from the buffer
+      RET_IF_RESULT_NEGATIVE(writer->write(output.digits[pos], 1));
```
the `write` function has an overload that just takes a char without a length, so anywhere you're doing `writer->write(char, 1)` can just be `writer->write(char)`

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+  return convert_inf_nan(writer, to_conv);
+}
+
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE int convert_float_decimal_typed(Writer *writer,
+                                            const FormatSection &to_conv,
+                                            fputil::FPBits<T> float_bits) {
+  return convert_float_typed<T>(writer, to_conv, float_bits, ConversionType::F);
+}
+
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE int convert_float_dec_exp_typed(Writer *writer,
+                                            const FormatSection &to_conv,
+                                            fputil::FPBits<T> float_bits) {
+  return convert_float_typed<T>(writer, to_conv, float_bits, ConversionType::E);
+}
+
+template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
+LIBC_INLINE int convert_float_dec_auto_typed(Writer *writer,
+                                             const FormatSection &to_conv,
+                                             fputil::FPBits<T> float_bits) {
+  return convert_float_typed<T>(writer, to_conv, float_bits, ConversionType::G);
+}
```
these are unused

statham-arm replied:
I don't think they are – `libc/src/stdlib/str_from_util.h` refers to them? Surely if the usual `float_dec_converter.h` provides them for that purpose, my replacement must provide them too.

michaelrj-google replied:
ah, I think I understand what happened. I tried deleting them and it still built because the `strfrom` code isn't getting the compile flags applied. For those to be affected you'd need to update the cmake in `src/stdlib`

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,593 @@
+
+#include <inttypes.h>
+#include <stddef.h>
+#include <string.h>
```
we try to avoid including public libc headers internally. For right now I'd recommend just using the `__builtin_mem***` variants since that's simplest.

michaelrj-google reviewed (commented):
This patch looks very good, I have some comments about the details of the algorithm but I like the direction you're going.

statham-arm committed:
fixup! [libc] Alternative algorithm for decimal FP printf

statham-arm commented:
Thanks for the comments! I've pushed a fixup patch dealing with the two easy ones (avoid including public headers, use the `write(char)` overload). I'll need to do some more rework for the larger issue of trying to avoid the loop.

statham-arm committed:
Substantial rewrite to eliminate the loop.

Compared to the previous version:

The helper function `need_to_round_up` is replaced with
`rounding_direction`, which has three rather than two return values:
it can distinguish 'round down' from 'already exact, no rounding at
all'.

The method `DyadicFloat::as_mantissa_type_rounded` no longer returns
an overflow indication, because we now avoid overflowing in the first
place by early detection of wildly out-of-range exponents. But it does
return a rounding direction indicating whether the mantissa was
rounded up, down, or was exact.

I had to fix `BigInt::decrement()`, which turned out to be
accidentally an exact copy of `increment()`.

Then `decimal_digits` is rewritten essentially as suggested in the
review comments:
 - there's no loop
 - the initial estimated exponent is used to detect the need to shift
   from F to E mode in advance rather than spotting it after the fact
   by overflow
 - so now the initial decimal exponent is always either correct or one
   too small
 - and if it's one too small, we handle it by truncating a digit off
   the output decimal mantissa, re-rounding carefully to avoid a
   double-rounding error.

statham-arm committed:
Build and format fixes

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,669 @@
+      // at a time. (A bit painful, but better than going back to the integer
+      // we made it from and doing the decimal conversion all over again.)
+      for (size_t i = output.ndigits; i-- > 0;) {
+        if (output.digits[i]++ != '9')
```
nit: same as above, I personally prefer to keep decrements/increments separate.

For character encoding independence this loop is fine, since it's only checking equality.

statham-arm replied:
I'm confused – surely here I'm depending on exactly the same assumption about character encoding that I did in the earlier `char_value - '0'` expression? Namely, that the decimal digit characters appear consecutively, so that I can increment one to the next digit by writing `output.digits[i]++`.

Neither piece of code depends on the _specific_ ASCII encodings of the decimal digits. But both depend on them being consecutive. So either you've misunderstood, or I have: why is one of those OK and the other not?

michaelrj-google replied:
you're right, I did misunderstand. This should be `output.digits[i] = int_to_b36_char(b36_char_to_int(output.digits[i]) + 1);`

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,669 @@
+                          radix::Dec::WithWidth<2>::WithSign>
+        expcvt{output.exponent};
+    cpp::string_view expview = expcvt.view();
+    expbuf[0] = (to_conv.conv_name & 32) | 'E';
```
For character encoding independence, use `isupper` and `toupper` here.

michaelrj-google reviewed (commented):
The algorithm looks good to me, I've left some comments but they're mostly code style related. Thank you for figuring out how to get rid of the loop, I think the result is much more elegant.

statham-arm committed:
Address review nits from 2025-01-27

statham-arm commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,669 @@
+    // Extract the two relevant digits. round_digit is the one we're removing;
+    // new_low_digit is the last one we're keeping, so we need to know if it's
+    // even or odd to handle exact tie cases (when round_dir == 0).
+    int round_digit = output.digits[--output.ndigits] - '0';
+    int new_low_digit =
+        output.ndigits == 0 ? 0 : output.digits[output.ndigits - 1] - '0';
```
Wow, that's surely above the call of duty? The C standard has always guaranteed that the `char` values of the decimal digits are consecutive. Not the alphabet, because EBCDIC was still a thing when C was getting started, but even EBCDIC puts 0,1,...,9 in sensible places.

michaelrj-google commented on libc/src/stdio/printf_core/float_dec_converter_limited.h:
```diff
@@ -0,0 +1,669 @@
+  // Convert the mantissa into a DyadicFloat, making sure it has the right
+  // sign, so that directed rounding will go in the right direction, if
+  // enabled.
+  fputil::DyadicFloat<DF_BITS> flt_mantissa(input.sign, input.exponent - 127,
+                                            input.mantissa);
```
in that case it might make sense to just subtract `<bits in mantissa> - 1` for clarity.

statham-arm committed:
Address review nits 2025-01-28

lntue commented on libc/src/__support/FPUtil/dyadic_float.h:
```diff
@@ -26,6 +26,49 @@
+      // We round up, unless the value is an exact halfway case and
+      // the bit that will end up in the units place is 0, in which
+      // case tie-break-to-even says round down.
+      return value.get_bit(rshift) != 0 || (value << (Bits - rshift + 1)) != 0
```
It looks like `value.get_bit(rshift)` will be overflow if `rshift == Bits`?

lntue commented on libc/src/__support/FPUtil/dyadic_float.h:
```diff
@@ -26,6 +26,49 @@
+  case FE_TOWARDZERO:
+    return -1;
+  case FE_DOWNWARD:
+    return logical_sign.is_neg() && (value << (Bits - rshift)) != 0 ? +1 : -1;
```
`Bits - rshift` could be negative here?

lntue commented on libc/src/__support/FPUtil/dyadic_float.h:
```diff
@@ -26,6 +26,49 @@
+  case FE_DOWNWARD:
+    return logical_sign.is_neg() && (value << (Bits - rshift)) != 0 ? +1 : -1;
+  case FE_UPWARD:
+    return logical_sign.is_pos() && (value << (Bits - rshift)) != 0 ? +1 : -1;
```
ditto

lntue commented on libc/src/__support/FPUtil/dyadic_float.h:
```diff
@@ -464,6 +567,96 @@ LIBC_INLINE constexpr DyadicFloat<Bits> quick_mul(const DyadicFloat<Bits> &a,
+  }
+
+  DyadicFloat<(Bits * 2)> qbig(qf.sign, qf.exponent - 2, q);
+  auto qfinal = DyadicFloat<Bits>::round(qbig.sign, qbig.exponent + Bits,
```
nit: just `return DyadicFloat<Bits>::round...`

lntue reviewed (commented):
Thanks @statham-arm !  It looks good to me with just a few nits.

statham-arm committed:
Address review nits 2025-01-29

michaelrj-google reviewed (approved):
LGTM, thanks for writing this patch!

statham-arm commented:
Thank you both for approving!

Would you please also look at #123580, which this FP printf strategy depends on? It _works_ correctly with or without that patch, but it's a huge slowdown without it, because this patch depends on `IntegerToString<BigInt>` to make its output digit strings, and that's currently much slower than it needs to be.
